### PR TITLE
Fix #53: Update Tests to probot v9 and fix all test issues

### LINKED ISSUE #53: Issue with running tests.
I investigate the test failures and I was able to fix some of them. Others are failing for legitimate reasons, which is why I created this issue.

1. The scheduler gets created at https://github.com/oppia/oppiabot/blob/6d723a95549353bb1641325f62efc95c8f420f64/index.js#L11, but in the tests, the app (oppiabot) is initialized https://github.com/oppia/oppiabot/blob/6d723a95549353bb1641325f62efc95c8f420f64/spec/oppiabotSpec.js#L16 before `robot.auth` gets mocked https://github.com/oppia/oppiabot/blob/6d723a95549353bb1641325f62efc95c8f420f64/spec/oppiabotSpec.js#L82 and the scheduler calls `robot.auth` https://github.com/probot/scheduler/blob/8ad237c69116894a5579d360e544f2586191fa7a/index.js#L77 before it is mocked so it calls the actual `robot.auth` and not the mocked implementation.

A possible fix for this would be to mock `robot.auth` before initializing oppiabot. I did that and it fixed the initial problem but provided another problem: the mock github data which is being resolved misses some data required by the scheduler.
2. Before every event gets dispatched, we check if the repo owner is allowed to run the bot https://github.com/oppia/oppiabot/blob/6d723a95549353bb1641325f62efc95c8f420f64/index.js#L23. Going through the mock payload, the repo owner is set to oppia, so adding oppia to the whitelisted account in the .env will make this work.

After fixing the above issues, 8 tests pass and 6 fail. 
3. This test https://github.com/oppia/oppiabot/blob/6d723a95549353bb1641325f62efc95c8f420f64/spec/oppiabotSpec.js#L103 checks that the function is called with 3 arguments, but the function is only called with one (context) which is expected since probot calls the handlers with just one argument. 
4. https://github.com/oppia/oppiabot/blob/6d723a95549353bb1641325f62efc95c8f420f64/spec/oppiabotSpec.js#L108 tests that the first argument passed is the username of the user, but it is instead the context , and again this is the expected behaviour.
Other tests fail for similar reason. maybe this is from an earlier implementation?

### PR DESCRIPTION
**Pull request by jameesjohn**
## Explanation

Fixes #53: Fixes the issue with running tests and upgrade the test to probot v9.
## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

**Comment by jameesjohn** [comment]
PTAL @ankita240796
Thanks

File: fixtures/pullRequestPayload.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "event": "pull_request",
+  "name": "pull_request",
```
Comment (ankita240796) [inline fixtures/pullRequestPayload.json:2]: Why this change?

File: spec/oppiabotSpec.js [hunk 2]
```diff
@@ -1,22 +1,29 @@
 require('dotenv').config();
-
-const { createRobot } = require('probot');
+const {createProbot} = require('probot');
```
Comment (ankita240796) [inline spec/oppiabotSpec.js:2]: Add spaces after braces: { createProbot }

File: spec/oppiabotSpec.js [hunk 3]
```diff
@@ -102,24 +119,18 @@ describe('Oppiabot\'s', () => {
...
+        expect(context.payload.pull_request.user.login).toEqual(
+          'testuser7777');
       });
 
-    it('should be called for a pull request for the given payload', () => {
-      expect(
-        apiForSheetsModule.checkClaStatus.calls.argsFor(0)[2])
-        .toEqual(true);
-    });
-
```
Comment (ankita240796) [inline spec/oppiabotSpec.js:122]: Why is this removed?

File: spec/oppiabotSpec.js [hunk 4]
```diff
@@ -102,24 +119,18 @@ describe('Oppiabot\'s', () => {
       expect(apiForSheetsModule.checkClaStatus.calls.count()).toEqual(1);
     });
 
-    it('should be called with three arguments for the given payload', () => {
+    it('should be called with one argument for the given payload', () => {
```
Comment (ankita240796) [inline spec/oppiabotSpec.js:122]: Can you please explain this change?

File: fixtures/pullRequestPayload.json (hunk 1 above)
Comment (jameesjohn) [reply fixtures/pullRequestPayload.json:2]: This change is made because probot expects to receive a webhook with an object of the structure 
`
{  
name: <event_type>, 
payload: <webhookpayload>
}
`
 https://github.com/oppia/oppiabot/blob/9eaf808aacfe8f8ff97ce46a9b7a0182a3616ae8/spec/oppiabotSpec.js#L110

File: spec/oppiabotSpec.js (hunk 2 above)
Comment (jameesjohn) [reply spec/oppiabotSpec.js]: Done.

File: spec/oppiabotSpec.js (hunk 4 above)
Comment (jameesjohn) [reply spec/oppiabotSpec.js:122]: `apiForSheetsModule.checkClaStatus` expects to be called with just one argument `Context`
https://github.com/oppia/oppiabot/blob/04d7f3b8723165bbbdc51074855d5223af75b2cf/index.js#L27

File: spec/oppiabotSpec.js (hunk 3 above)
Comment (jameesjohn) [reply spec/oppiabotSpec.js:122]: I'm not sure what was being tested here because `apiForSheetsModule` expects just one argument, and the check is testing for the third argument.

**Comment by jameesjohn** [comment]
Hi @ankita240796, I've responded to comments, PTAL. Thanks!

File: spec/oppiabotSpec.js (hunk 3 above)
Comment (jameesjohn) [reply spec/oppiabotSpec.js:122]: I'm not sure what we should test here.

**Comment by jameesjohn** [comment]
> Regarding running tests locally, I think we should have clear instructions and a python script... 


@ankita240796, would it be okay if I write the script in node?

File: spec/oppiabotSpec.js (hunk 3 above)
Comment (ankita240796) [reply spec/oppiabotSpec.js:122]: I think the test is related to contextData, can you try to figure out what was being compared with true in the original test.

File: fixtures/pullRequestPayload.json (hunk 1 above)
Comment (ankita240796) [reply fixtures/pullRequestPayload.json:2]: Ok, sg.

File: spec/oppiabotSpec.js (hunk 4 above)
Comment (ankita240796) [reply spec/oppiabotSpec.js:122]: Ok, sg.

**Comment by ankita240796** [comment]
> > Regarding running tests locally, I think we should have clear instructions and a python script...
> 
> @ankita240796, would it be okay if I write the script in node?

Yes, any language is fine as long as we have clear instructions on how to run.

**Comment by jameesjohn** [comment]
PTAL @ankita240796
I've created a `setup.js` file which will handle updating the `.env`, running the tests and reverting it back once the tests are done running.

File: spec/oppiabotSpec.js (hunk 3 above)
Comment (jameesjohn) [reply spec/oppiabotSpec.js:122]: @ankita240796 I think it checks that the data gotten is a pull request data
https://github.com/oppia/oppiabot/blob/3aa31ef80b5af0789e34d608ef85ec9a11becd61/index.js#L32
https://github.com/oppia/oppiabot/blob/3aa31ef80b5af0789e34d608ef85ec9a11becd61/lib/apiForSheets.js#L114

**Comment by jameesjohn** [comment]
PTAL @ankita240796. I've responded to all comments.

File: lib/apiForSheets.js [hunk 5]
```diff
@@ -188,6 +188,8 @@ module.exports.checkClaStatus = function(contextData) {
 
   // Authorize a client with the loaded credentials, then call the
   // Google Sheets API.
-  module.exports.authorize(
+  const oauthClient = module.exports.authorize(
     JSON.parse(clientSecret), module.exports.checkClaSheet);
+
+  module.exports.checkClaSheet(oauthClient);
 };
```
Comment (ankita240796) [inline lib/apiForSheets.js:195]: Can you explain the purpose of this change i.e. why is it required for running the tests?

File: spec/oppiabotSpec.js (hunk 3 above)
Comment (ankita240796) [reply spec/oppiabotSpec.js:122]: So, we should test the same in the current tests as well.

File: spec/setup.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
+/* eslint-disable no-console */
+const fs = require('fs');
```
Comment (ankita240796) [inline spec/setup.js:2]: Let us add instructions on how to run this as well as a copyright license here and everywhere else.

Also, can you provide a set of points on what you have tried and what are the issues you are facing with newrelic while running the tests (You can create a doc as well).

File: spec/setup.js [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+const path = require('path');
+const { EOL } = require('os');
+const { exec } = require('child_process');
+
+const envPath = path.join(__dirname, '..', '.env');
+let envData = '';
+
+const setWhitelistedAccount = () => {
+  // Load env
```
Comment (ankita240796) [inline spec/setup.js:11]: End in period.

File: spec/setup.js [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  const data = fs.readFileSync(envPath, (err) => {
+    if (err) {
+      throw err;
+    }
+  });
+  envData = data.toString();
+
+  // Parse and remove comments
```
Comment (ankita240796) [inline spec/setup.js:19]: End in period.

File: spec/setup.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  let envArray = data.toString().split(EOL);
+  envArray = envArray.filter((line) => {
+    if (line && !line.startsWith('#')) {
+      return line;
+    }
+  });
+
+  // check for whtelisted_account
```
Comment (ankita240796) [inline spec/setup.js:27]: Capitalize first letter, end in period.

File: spec/setup.js [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  const whitelist = envArray.find((line) =>
+    line.startsWith('WHITELISTED_ACCOUNTS')
+  );
+  const whitelistIndex = envArray.indexOf(whitelist);
+  if (!whitelist.includes('oppia')) {
+    // Replace the current whitelist with new one
```
Comment (ankita240796) [inline spec/setup.js:33]: End in period.

File: spec/setup.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  let envArray = data.toString().split(EOL);
+  envArray = envArray.filter((line) => {
+    if (line && !line.startsWith('#')) {
+      return line;
+    }
+  });
+
+  // check for whtelisted_account
+  const whitelist = envArray.find((line) =>
+    line.startsWith('WHITELISTED_ACCOUNTS')
```
Comment (ankita240796) [inline spec/setup.js:29]: `WHITELISTED_ACCOUNTS`: Declare this as a constant and use everywhere.

File: spec/setup.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  const whitelistIndex = envArray.indexOf(whitelist);
+  if (!whitelist.includes('oppia')) {
+    // Replace the current whitelist with new one
+    const newWhitelist = 'WHITELISTED_ACCOUNTS=oppia';
+    envArray.splice(whitelistIndex, 1, newWhitelist);
+  }
+
+  // set client secret
+  const clientSecretIndex = envArray.findIndex((line) =>
+    line.startsWith('CLIENT_SECRET')
```
Comment (ankita240796) [inline spec/setup.js:40]: Ditto.

File: spec/setup.js [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+const runTest = () => {
+  const jasminePath = path.join(
+    __dirname,
+    '..',
+    'node_modules',
+    '.bin',
+    'jasmine'
+  );
```
Comment (ankita240796) [inline spec/setup.js:62]: Move in single line.

File: spec/oppiabotSpec.js (hunk 3 above)
Comment (jameesjohn) [reply spec/oppiabotSpec.js:122]: I don't think it is necessary to check if it is a pull request data or not since we are only calling the `apiForSheetsModule.checkClaStatus` function when a pull request is opened.
I believe this was done initially because the checkClaStatus was also called when an issue comment was created.
https://github.com/oppia/oppiabot/blob/511f4f06332f73b2efa9a258305d1821d3aaacac/index.js#L15

File: spec/setup.js (hunk 6 above)
Comment (jameesjohn) [reply spec/setup.js:2]: This can be run by running `npm run test`.
I added it to the test script in package.json https://github.com/oppia/oppiabot/blob/6c02cbd51d7bec70a8ced69fa26770c69ebaae16/package.json#L12

File: spec/setup.js (hunk 7 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

File: spec/setup.js (hunk 8 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

File: spec/setup.js (hunk 9 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

File: spec/setup.js (hunk 13 above)
Comment (jameesjohn) [reply spec/setup.js:65]: Keeping them all on one line is longer than 80 lines and eslint complains.

File: spec/setup.js (hunk 11 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

File: spec/setup.js (hunk 10 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

File: spec/setup.js (hunk 12 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

**Comment by jameesjohn** [comment]
I've responded to all comments, PTAL @ankita240796. Thanks!

File: lib/apiForSheets.js (hunk 5 above)
Comment (ankita240796) [reply lib/apiForSheets.js]: Thanks for the explanation!

File: spec/setup.js (hunk 13 above)
Comment (ankita240796) [reply spec/setup.js:83]: You can split into two lines.

File: spec/oppiabotSpec.js (hunk 3 above)
Comment (ankita240796) [reply spec/oppiabotSpec.js:122]: Ok, sg.

File: spec/setup.js (hunk 6 above)
Comment (ankita240796) [reply spec/setup.js:20]: Hmm, so the test is written in a way that considers `.env` is already present in the oppiabot folder. Let us update it so that it creates `.env` using `.env.example` if it is missing.

**Comment by ankita240796** [comment]
Hey @jameesjohn, left a few more comments, PTAL. Thanks!

File: spec/setup.js (hunk 13 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

File: spec/setup.js (hunk 6 above)
Comment (jameesjohn) [reply spec/setup.js:20]: Done.

**Comment by jameesjohn** [comment]
Done @ankita240796, PTAL. Thanks!

**Comment by ankita240796** [comment]
Hmm ok, so I see that the tests fail since newrelic config is not found. Is this the same issue you are facing?

**Comment by jameesjohn** [comment]
No
Mine passes without the new relic config.
Can you send a screenshot?

**Comment by ankita240796** [comment]
Here is the log:

```
Updating .env
Error: Command failed: "/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/.bin/jasmine"
Unable to find New Relic module configuration. A default
configuration file can be copied from /Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/newrelic/lib/config/default.js
and put at /Users/saxenaankita/Desktop/opensrc/oppiabot/newrelic.js or
/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/jasmine/bin/newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true
Unable to find New Relic module configuration. A default
configuration file can be copied from /Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/newrelic/lib/config/default.js
and put at /Users/saxenaankita/Desktop/opensrc/oppiabot/newrelic.js or
/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/jasmine/bin/newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true

    at ChildProcess.exithandler (child_process.js:303:12)
    at ChildProcess.emit (events.js:311:20)
    at maybeClose (internal/child_process.js:1021:16)
    at Socket.<anonymous> (internal/child_process.js:443:11)
    at Socket.emit (events.js:311:20)
    at Pipe.<anonymous> (net.js:668:12) {
  killed: false,
  code: 1,
  signal: null,
  cmd: '"/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/.bin/jasmine"'
}
Unable to find New Relic module configuration. A default
configuration file can be copied from /Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/newrelic/lib/config/default.js
and put at /Users/saxenaankita/Desktop/opensrc/oppiabot/newrelic.js or
/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/jasmine/bin/newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true
Unable to find New Relic module configuration. A default
configuration file can be copied from /Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/newrelic/lib/config/default.js
and put at /Users/saxenaankita/Desktop/opensrc/oppiabot/newrelic.js or
/Users/saxenaankita/Desktop/opensrc/oppiabot/node_modules/jasmine/bin/newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true

Started
{"v":0,"level":30,"name":"newrelic_bootstrap","hostname":"saxenaankita-macbookpro.roam.corp.google.com","pid":23333,"time":"2020-05-04T16:38:08.166Z","msg":"Using New Relic for Node.js. Agent version: 4.13.1; Node version: v12.16.1."}
{"v":0,"level":40,"name":"newrelic_bootstrap","hostname":"saxenaankita-macbookpro.roam.corp.google.com","pid":23333,"time":"2020-05-04T16:38:08.167Z","msg":"New Relic for Node.js 4.13.1 has not been tested on Node.js v12.16.1. Please update the agent or downgrade your version of Node.js"}
{"v":0,"level":30,"name":"newrelic_bootstrap","hostname":"saxenaankita-macbookpro.roam.corp.google.com","pid":23333,"time":"2020-05-04T16:38:08.167Z","msg":"Module not enabled in configuration; not starting."}
FFFFFFFFFFFFF

Failures:
1) Oppiabot's apiForSheets should be called for the given payload
  Message:
    TypeError: createProbot is not a function
  Stack:
        at <Jasmine>
        at UserContext.<anonymous> (/Users/saxenaankita/Desktop/opensrc/oppiabot/spec/oppiabotSpec.js:93:13)
        at <Jasmine>
        at processImmediate (internal/timers.js:456:21)
        at process.topLevelDomainCallback (domain.js:137:15)
  Message:
    TypeError: Cannot read property 'receive' of undefined
  Stack:
        at <Jasmine>
        at UserContext.<anonymous> (/Users/saxenaankita/Desktop/opensrc/oppiabot/spec/oppiabotSpec.js:110:13)
        at <Jasmine>
        at processImmediate (internal/timers.js:456:21)
        at process.topLevelDomainCallback (domain.js:137:15)
  Message:
    Expected spy checkClaStatus to have been called.
  Stack:
    Error: Expected spy checkClaStatus to have been called.
        at <Jasmine>
        at UserContext.<anonymous> (/Users/saxenaankita/Desktop/opensrc/oppiabot/spec/oppiabotSpec.js:115:49)
        at <Jasmine>
        at processImmediate (internal/timers.js:456:21)
        at process.topLevelDomainCallback (domain.js:137:15)
```

**Comment by jameesjohn** [comment]
Uhmm...
Which version of node.js are you using?

When I run the test, it gives the new relic error message, but it continues to run the tests.

**Comment by jameesjohn** [comment]
```
Updating .env
Unable to find New Relic module configuration. A default
configuration file can be copied from C:\Users\PT WORLD\opensource\oppiabot\node_modules\newrelic\lib\config\default.js
and put at C:\Users\PT WORLD\opensource\oppiabot\newrelic.js or
C:\Users\PT WORLD\opensource\oppiabot\node_modules\jasmine\bin\newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true
Unable to find New Relic module configuration. A default
configuration file can be copied from C:\Users\PT WORLD\opensource\oppiabot\node_modules\newrelic\lib\config\default.js
and put at C:\Users\PT WORLD\opensource\oppiabot\newrelic.js or
C:\Users\PT WORLD\opensource\oppiabot\node_modules\jasmine\bin\newrelic.js. If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true

Started
{"v":0,"level":30,"name":"newrelic_bootstrap","hostname":"JamesSpectre","pid":14208,"time":"2020-05-04T16:28:25.871Z","msg":"Using New Relic for Node.js. Agent version: 4.13.1; Node version: v10.16.3."}
{"v":0,"level":30,"name":"newrelic_bootstrap","hostname":"JamesSpectre","pid":14208,"time":"2020-05-04T16:28:25.873Z","msg":"Module not enabled in configuration; not starting."}
16:28:27.587Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.602Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.636Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.648Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.707Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.723Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.754Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.768Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.782Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.812Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.828Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.858Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.16:28:27.882Z DEBUG probot: Webhook received
  event: {
    "event": "pull_request.opened",
    "installation": 213174,
    "repository": "oppia/oppia"
  }
RUNNING CHANGELOG LABEL CHECK ON PULL REQUEST 5139 ...
.


13 specs, 0 failures
Finished in 0.596 seconds

Reverting .env
```

**Comment by ankita240796** [comment]
Ok, so let us do two things:

1. Let us set this var `NEW_RELIC_NO_CONFIG_FILE` to true in env based on the message here: 
```
If you are not using file based config
please set the environment variable NEW_RELIC_NO_CONFIG_FILE=true
```

2. Let us have a script which checks and installs the required versions of all the deps before running the tests (as we have in oppia repo).

File: lib/apiForSheets.js [hunk 14]
```diff
@@ -188,6 +188,7 @@ module.exports.checkClaStatus = function(contextData) {
 
   // Authorize a client with the loaded credentials, then call the
   // Google Sheets API.
-  module.exports.authorize(
-    JSON.parse(clientSecret), module.exports.checkClaSheet);
+  const oauthClient = module.exports.authorize(JSON.parse(clientSecret));
+
+  module.exports.checkClaSheet(oauthClient);
```
Comment (ankita240796) [inline lib/apiForSheets.js:193]: Let us add an explanation here as well.

File: package.json [hunk 15]
```diff
@@ -9,13 +9,13 @@
     "dev": "nodemon",
     "lint": "standard",
     "start": "probot run ./index.js",
-    "test": "./node_modules/.bin/jasmine"
+    "test": "node ./spec/setup.js"
```
Comment (ankita240796) [inline package.json:12]: Can we do this change for package.json in a separate PR?

File: lib/apiForSheets.js (hunk 14 above)
Comment (jameesjohn) [reply lib/apiForSheets.js]: Done

**Comment by jameesjohn** [comment]
> Hi @jameesjohn, I am still not able to run the tests

Hi @ankita240796, did you run `npm install`? Since newrelic got updated, I think you need to reinstall it.

> why do we specifically need to remove package-lock.json from .gitignore?

We need package-lock.json committed so that we can ensure that the same version of dependencies are installed by all the contributors.

**Comment by ankita240796** [comment]
Hi @jameesjohn, yes it works fine now, I missed updating by running `npm install`. Taking a final pass now. Thanks!

File: spec/setup.js [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+    line.startsWith(CLIENT_SECRET)
+  );
+  const newClientSecret = CLIENT_SECRET + '="{}"';
+  envArray.splice(clientSecretIndex, 1, newClientSecret);
+
+  // Update new relic config.
+  const newRelicConfigIndex = envArray.findIndex((line) =>
+    line.startsWith(NEW_RELIC)
+  );
+  const newRelicConfig = 'NEW_RELIC_NO_CONFIG_FILE=true';
```
Comment (ankita240796) [inline spec/setup.js:74]: Interpolate the constant here.

File: spec/setup.js [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+  if (newRelicConfigIndex !== -1) {
+    envArray.splice(newRelicConfigIndex, 1, newRelicConfig);
+  } else {
+    envArray.push(newRelicConfig);
+  }
+
+  const newRelicAppIndex = envArray.findIndex((line) =>
+    line.startsWith(NEW_RELIC_APP)
+  );
+  const newRelicConfigApp = 'NEW_RELIC_APP_NAME=oppiabot';
```
Comment (ankita240796) [inline spec/setup.js:84]: Ditto.

File: spec/setup.js [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+  console.log('Updating .env');
+  fs.writeFileSync(envPath, newEnv, (err) => {
+    if (err) {
+      throw err;
+    }
+  });
+};
+
+const runTest = () => {
+  const jasminePath = path.join(__dirname, '..', 'node_modules',
```
Comment (ankita240796) [inline spec/setup.js:102]: Break after opening bracket.

File: spec/setup.js [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+const envExamplePath = path.join(__dirname, '..', '.env.example');
+let envData = '';
+
+const setWhitelistedAccount = () => {
+  // Load env.
+  let data = '';
+  if (fs.existsSync(envPath)) {
+    data = fs.readFileSync(envPath);
+  } else {
+    data = fs.readFileSync(envExamplePath);
```
Comment (ankita240796) [inline spec/setup.js:39]: Have a check here: if example file is missing as well, throw an error asking the user to update their branch.

Comment (jameesjohn) [reply spec/setup.js:39]: Done.

File: spec/setup.js (hunk 16 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

File: spec/setup.js (hunk 17 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

File: spec/setup.js (hunk 18 above)
Comment (jameesjohn) [reply spec/setup.js]: Done.

**Comment by jameesjohn** [comment]
Hi @ankita240796, I've addressed all the review comments, PTAL. Thanks!

> We should set up coverage check as well either in this PR or a separate one

Yes, I agree. I think it should be in a separate PR.

File: spec/setup.js [hunk 20]
```diff
@@ -35,8 +35,11 @@ const setWhitelistedAccount = () => {
   let data = '';
   if (fs.existsSync(envPath)) {
     data = fs.readFileSync(envPath);
-  } else {
+  } else if (fs.existsSync(envExamplePath)) {
     data = fs.readFileSync(envExamplePath);
+  } else {
+    throw new Error(
+      '.env.example could not be found. Please update your branch with master');
```
Comment (ankita240796) [inline spec/setup.js:42]: End with Period.

Comment (jameesjohn) [reply spec/setup.js]: Done.
